baseapp/state.go

```diff
@@ -17,5 +20,15 @@ func (st *state) CacheMultiStore() sdk.CacheMultiStore {
 
 // Context returns the Context of the state.
 func (st *state) Context() sdk.Context {
+	defer st.lock.RUnlock()
+	st.lock.RLock()
+
 	return st.ctx
 }
+
+// WithContext update context of the state
+func (st *state) WithContext(ctx sdk.Context) {
```

Comment:
`WithX` methods typically leave the receiver unmodified and return a copy with the requested changes. Should this be e.g. `SetContext`?

Reply:
💯